### PR TITLE
Add initial CAS protocols

### PR DESCRIPTION
**Pull request by dmbryson**

### CODE REVIEW

File: Sources/llbuild2/CAS/Object.swift [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+        var container = encoder.container(keyedBy: CodingKeys.self)
+        try container.encode(refs, forKey: .refs)
+        try container.encode(data.getBytes(at: 0, length: data.readableBytes), forKey: .data)
+    }
+
+    public init(from decoder: Decoder) throws {
+        let values = try decoder.container(keyedBy: CodingKeys.self)
+        self.refs = try values.decode([LLBDataID].self, forKey: .refs)
+
+        let bytes = try values.decode([UInt8].self, forKey: .data)
```
Comment (dmbryson) [inline Sources/llbuild2/CAS/Object.swift:52]: @vlm I don't think I'm holding this right.   Seems like there should be a more direct means of decoding into the ByteBuffer here.   Thoughts?

File: Sources/llbuild2/CAS/DataID.swift [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+// See http://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+
+import Foundation
+
+
+/// An identifier that uniquely identifies a CAS object (i.e. a hash).
+public struct LLBDataID: Hashable, CustomDebugStringConvertible {
+    private enum IDKind: UInt8 {
+    /// An id that is directly calculated based on a hash of the data.
+    case directHash = 0
```
Comment (sergiocampama) [inline Sources/llbuild2/CAS/DataID.swift:16]: nit: indent?

File: Sources/llbuild2/CAS/DataID.swift [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+}
+
+
+extension LLBDataID: Comparable {
+    /// Compare DataID according to stable but arbitrary rules
+    /// (not necessarily alphanumeric).
+    public static func < (lhs: LLBDataID, rhs: LLBDataID) -> Bool {
+        let a = lhs.bytes
+        let b = rhs.bytes
+        if a.count == b.count {
```
Comment (sergiocampama) [inline Sources/llbuild2/CAS/DataID.swift:80]: do you also need to check for the kind here?

File: Sources/llbuild2/CAS/DataID.swift [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+    public func encode(to encoder: Encoder) throws {
+        var container = encoder.singleValueContainer()
+        try container.encode(self.debugDescription)
+    }
+}
+
+
+extension LLBDataID {
+    public enum LLBDataIDSliceError: Error {
+    case decoding(String)
```
Comment (sergiocampama) [inline Sources/llbuild2/CAS/DataID.swift:116]: nit: indent?

File: Sources/llbuild2/CAS/Database.swift [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+
+import Foundation
+
+import NIO
+
+
+/// Features supported by a CAS Database
+public struct LLBCASFeatures: Codable {
+    
+    /// Whether a database is "ID perserving"
```
Comment (sergiocampama) [inline Sources/llbuild2/CAS/Database.swift:17]: nit: preserving

File: Sources/llbuild2/CAS/Database.swift [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+    }
+}
+
+/// A content-addressable database protocol
+///
+/// THREAD-SAFETY: The database is expected to be thread-safe.
+public protocol LLBCASDatabase: class {
+    
+    /// Get the supported features of this database implementation
+    func supportedFeatures() -> EventLoopFuture<LLBCASFeatures>
```
Comment (sergiocampama) [inline Sources/llbuild2/CAS/Database.swift:35]: I was wondering if we should have an LLBSupport module with shorter typealiases like `Future`. We will also need a bunch of things in support anyways, so we might as well add it at some point (not for this PR though)

File: Sources/llbuild2/CAS/Object.swift [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+    var size: Int {
+        return data.readableBytes
+    }
+}
+
+// MARK: Codable support for CASObject
+
+extension LLBCASObject: Codable {
+    private enum CodingKeys: String, CodingKey {
+    case refs
```
Comment (sergiocampama) [inline Sources/llbuild2/CAS/Object.swift:38]: nit: indent?

File: Sources/llbuild2/CAS/Database.swift (hunk 6 above)
Comment (dmbryson) [reply Sources/llbuild2/CAS/Database.swift:35]: I indeed intend to follow up with a typealias.  These are getting a bit long.

File: Sources/llbuild2/CAS/Object.swift (hunk 7 above)
Comment (dmbryson) [reply Sources/llbuild2/CAS/Object.swift:38]: But this is what Xcode wants the indent to be... 😉

File: Sources/llbuild2/CAS/DataID.swift (hunk 3 above)
Comment (dmbryson) [reply Sources/llbuild2/CAS/DataID.swift:80]: Kind is actually in bytes[0] too.

File: Sources/llbuild2/CAS/Database.swift (hunk 5 above)
Comment (dmbryson) [reply Sources/llbuild2/CAS/Database.swift:17]: d'oh

**Pull request merged**